aquifer/aquifer
# Extension System

This is a very large PR. It contains a comprehensive extension system that allows users to do everything described in this document: https://github.com/aquifer/aquifer/wiki/Adding,-removing,-loading,-and-listing-extensions
## NOTE

This should NOT be reviewed or merged before #14.
## Steps to test
- Checkout this branch.
- Run `npm install && npm link` within the aquifer repository.
- Checkout [this](https://github.com/aquifer/aquifer-drush) git repository in another folder.
- Create an aquifer project in any directory (not within this repo): `aquifer create myDir`
- Change your pwd into the directory aquifer should have created: `cd myDir`.
- Run `aquifer extension-add aquifer-drush -s /path/to/aquifer-drush`
- Ensure that aquifer returns with a success message about installing aquifer-drush.
- Run `aquifer extension-remove aquifer-drush`
- Ensure that aquifer returns a success message about removing aquifer-drush.
- Run `aquifer extension-add aquifer-drush -s git+ssh://git@github.com/aquifer/aquifer-drush` (re-install the module via github repo)
- Within your aquifer project dir, run `rm -r .aquifer/node_modules`
- Run `aquifer`
- Ensure that aquifer notifies you of the fact that your installed extensions are not downloaded.
- As the prompt should instruct, run `aquifer extensions-load`
- Within your aquifer project dir, run `ls .aquifer/node_modules`, and ensure that `aquifer-drush` is indeed loaded.
- Run `aquifer build`.
- Run `aquifer drush status`, and ensure that it returns valid information about an uninstalled Drupal site (the one you just built).
## Extension CLI help:

![screen shot 2015-06-28 at 5 50 24 pm](https://cloud.githubusercontent.com/assets/1107871/8400523/98512bf6-1de5-11e5-9b3f-8e78ec58f339.png)
## Extension installation:

![screen shot 2015-06-28 at 10 30 55 pm](https://cloud.githubusercontent.com/assets/1107871/8400531/be82a944-1de5-11e5-9956-8352fb8f3bfe.png)
## Extension uninstallation:

![screen shot 2015-06-28 at 10 30 37 pm](https://cloud.githubusercontent.com/assets/1107871/8400533/c70cacd6-1de5-11e5-8dda-509cdae374c6.png)
## Extensions out of sync, and being reloaded:

![screen shot 2015-06-28 at 10 31 43 pm](https://cloud.githubusercontent.com/assets/1107871/8400534/cf52c86c-1de5-11e5-8d10-e6d275230f1f.png)
